pallets/parachain-staking/src/types.rs

```diff
@@ -1362,6 +1382,48 @@ impl<
 		Err(Error::<T>::DelegationDNE.into())
 	}
 
+	/// Updates the bond locks for this delegator.
+	///
+	/// This will take the current self.total and ensure that a lock of the same amount is applied
+	/// and will also ensure that the account has at least `additional_reserve` in free balance.
+	///
+	/// `additional_reserve` should reflect the change to the amount that should be locked if
+	/// positive, 0 otherwise (e.g. `min(0, change_in_total_bond)`). This is necessary because it is
+	/// not possible to query the amount that is locked for a given lock id.
+	pub fn adjust_bond_lock<T: Config>(
+		&mut self,
+		additional_reserve: Option<Balance>,
+	) -> DispatchResult
+	where
+		BalanceOf<T>: From<Balance>,
+		T::AccountId: From<AccountId>,
+	{
+		if let Some(reserve) = additional_reserve {
+			ensure!(
+				<Pallet<T>>::get_delegator_stakable_free_balance(&self.id.clone().into())
+					>= reserve.into(),
+				Error::<T>::InsufficientBalance,
+			);
+
+			// additional sanity check: shouldn't ever want to reserve more than total
+			if reserve < self.total {
+				log::warn!("LOGIC ERROR: request to reserve more than bond total");
+			}
+		}
+
+		if self.total.is_zero() {
```

Comment:
This is a great concern, and one I've shared from the beginning of my work here. I've already uncovered a couple bugs where `self.total` had not yet been updated and the resulting lock was incorrect.

Compounding this problem is the potential for this struct to go out of sync with what's in storage, which I've encountered in at least one case.

I considered something similar to your `update_total` suggestion, but there are cases where this would result in multiple updates in a single extrinsic which would add a lot of wasted effort modifying locks.